### ISSUE TITLE
Test flicker on Phoenix build?

### ISSUE THREAD

**Issue by Trevoke** [phoenixframework/phoenix#2023]
I'm running the test suite while I'm preparing a PR, and this test just failed, once (test suite did not show same failure when I reran the suite right after):

```
  1) test warns on unhandled handle_info/2 messages (Phoenix.Test.ChannelTest)
     test/phoenix/test/channel_test.exs:414
     No message matching %Phoenix.Socket.Reply{ref: ^ref, status: :ok, payload: %{}} after 100ms.
     The following variables were pinned:
       ref = #Reference<0.0.2.8003>
     The process mailbox is empty.
     stacktrace:
       test/phoenix/test/channel_test.exs:419: anonymous fn/1 in Phoenix.Test.ChannelTest.test warns on unhandled handle_info/2 messages/1
       (ex_unit) lib/ex_unit/capture_log.ex:75: ExUnit.CaptureLog.capture_log/2
       test/phoenix/test/channel_test.exs:416: (test)

     The following output was logged:

     21:49:54.717 [error] Instrumenter Phoenix.Logger.phoenix_channel_join/3 failed.
     ** (CaseClauseError) no case clause matching: :error
         (phoenix) lib/phoenix/logger.ex:94: Phoenix.Logger.channel_log/3
         test/phoenix/test/channel_test.exs:13: Phoenix.Test.ChannelTest.Endpoint.instrument/4
         (phoenix) lib/phoenix/test/channel_test.ex:330: Phoenix.ChannelTest.join/4
         (phoenix) lib/phoenix/test/channel_test.ex:270: Phoenix.ChannelTest.subscribe_and_join!/4
         test/phoenix/test/channel_test.exs:415: Phoenix.Test.ChannelTest."test warns on unhandled handle_info/2 messages"/1
         (ex_unit) lib/ex_unit/runner.ex:296: ExUnit.Runner.exec_test/1
         (stdlib) timer.erl:166: :timer.tc/1
         (ex_unit) lib/ex_unit/runner.ex:246: anonymous fn/3 in ExUnit.Runner.spawn_test/3


     21:49:54.872 [error] Phoenix.Channel.Server #PID<0.1565.0> received unexpected message in handle_info/2: :unhandled
```

The latest commit Phoenix commit I have is df018f4 .

This is Elixir 1.3.4 .

**Comment by durub** [comment]
Would you have the seed the test suite was run with?

```
$ mix test
...............................................................

Finished in 0.7 seconds
63 tests, 0 failures

Randomized with seed 117662
```
In this case, 117662 was used to determine the order the tests were run. With it, we can replicate the specific test order used (which could trigger intermittent bugs), by running mix test -seed 117662.

If that test failed just once and the failure stems from the test order, you'd need the seed to replicate it.

**Comment by chrismccord** [comment]
I'm unable to recreate this on Elixir 1.3.4. Are you able to reliably hit this error after multiple runs?

**Comment by chrismccord** [comment]
It's also really unusual with the captured log output:

>      21:49:54.717 [error] Instrumenter Phoenix.Logger.phoenix_channel_join/3 failed.
     ** (CaseClauseError) no case clause matching: :error

This shouldn't be possible since the Channel module will have necessarily had the key/val that we are calling Map.fetch on.

**Comment by Trevoke** [comment]
The seed was 471806 -- but I can't reproduce it either.

My guess is that it's a timeout because whatever my box was doing in the
background made the test take longer than 100ms to process and reply.. In
conditions where it's the only thing it's doing, it behaves as it should.

I thought I should bring it up just in case it could be a concern.

**Comment by chrismccord** [comment]
Thanks for the report. We can chase down other causes, but I think you're right that you tripped the 100ms receive, perhaps your box was really cpu bound. Thanks!

**Comment by dustinfarris** [comment]
@chrismccord I am seeing this as well, when I run `mix test.watch`

Tests will pass, then if I write to a file (without changing anything) to trigger the watch, tests rerun and i get a failure with similar error as above:

```
The process mailbox is empty.
```

I am also on Elixir 1.3.4

**Comment by dustinfarris** [comment]
Doesn't seem to happen until there's been an actual error that is corrected

Repro:

Run `mix test.watch`

All tests pass.

Break something.. like a syntax error missing curly brace or something

Test fails.

Fix the syntax error

Tests pass (maybe)

From this point forward.  Saving files and triggering test re-runs, a channel test may sporadically fail with the "process mailbox is empty" message.

**Comment by dustinfarris** [comment]
@chrismccord can this be reopened?  i am seeing it a lot now

**Comment by chrismccord** [comment]
You need to explicitly provide a timeout if you are missing the 100ms wait time

**Comment by chrismccord** [comment]
to be clear, the fact it is timing out is not something to fix on our end but rather a product of your app+hardware, so you need to deviate from the 100ms defaults with `assert_` and friends

**Comment by dustinfarris** [comment]
@chrismccord i am testing a very simple channel on a 2012 27-inch iMac 3.4 GHz i7 32GB RAM

I am trying to triage further, but I don't think this is a hardware issue.

**Comment by josevalim** [comment]
@dustinfarris do you have an app that reproduces the issue? If this is an error happening in your app, we shouldn't assume it is immediately related to this issue. It is also worth trying to increase the timeout anyway and see if that fixes the issue, because if it does, then it is timeout related.

**Comment by dustinfarris** [comment]
@josevalim yeah increasing the timeout seems to make it go away.

it happens on the first channel test that runs, and only if the tests have not run in a while.  so i'm wondering if it is the initial postgresql connection that is causing it

**Comment by josevalim** [comment]
@dustinfarris that may be it. You can validate it by doing something like SELECT 1 in your test helper.

**Comment by dustinfarris** [comment]
no, i guess that isn't it :-/

**Comment by chrismccord** [comment]
Does your channel code call into any kind of lazy code path? Something that needs to spin up a pool of connections? Open and cache files? Etc?

**Comment by dustinfarris** [comment]
I have been able to reproduce this and believe it stems from absinthe_graphql.  Whether that is something that can be "fixed" remains to be seen.

I have opened a PR to absinthe's phoenix_chat_example demonstrating how this occurs:

https://github.com/benwilson512/phoenix_chat_example/pull/2

tl;dr it only happens on the first test, all tests afterward do not time out.

@chrismccord my code does not do anything like that, but possible something in absinthe does.
